### PR TITLE
[BugFix][Ming-TTS] Use uploaded direct speaker embeddings by voice

### PR DESCRIPTION
**Pull request by LHXuuu**
## Purpose

Fix Ming-TTS uploaded voice reuse when the voice was registered with a direct `speaker_embedding`.

Previously, Ming-TTS only tried to resolve uploaded voices as reference audio, so voices registered via `speaker_embedding` were not actually consumed by the Ming adapter. This PR loads direct uploaded speaker embeddings into `request.speaker_embedding`, while keeping the existing uploaded audio/ref_text path unchanged.

## Test Plan

**vLLM Version:** 0.24.0

**vLLM-Omni Commit:** aff21312

## Test Result

- Start Ming-TTS service:

```bash

vllm serve /home/admin/model \
  --host 0.0.0.0 \
  --port 8000 \
  --omni
```

Health check:

```bash
curl http://127.0.0.1:8000/health
```

- Generate a 192-dim Ming speaker embedding from reference audio:

```bash
python -c '
import json
from vllm_omni.model_executor.models.ming_tts.speaker_extractor import MingSpeakerEmbeddingExtractor

extractor = MingSpeakerEmbeddingExtractor("/home/admin/model", target_sr=16000)
emb = extractor.extract_from_file("/home/ming_stream_main.wav").detach().reshape(-1).float().cpu()
assert emb.numel() == 192, emb.numel()
open("/home/ming_stream_main_spk_emb_192.json", "w").write(json.dumps(emb.tolist()))
print("wrote /home/ming_stream_main_spk_emb_192.json")
'
```

Upload the 192-dim direct speaker embedding:

```bash
curl http://127.0.0.1:8000/v1/audio/voices \
  -F "name=upload_emb_voice_verify" \
  -F "consent=local_smoke_test" \
  -F "speaker_embedding=</home/ming_stream_main_spk_emb_192.json"
```

Result:

```json
{
  "success": true,
  "voice": {
    "name": "upload_emb_voice_verify",
    "consent": "local_smoke_test",
    "embedding_source": "direct",
    "embedding_dim": 192
  }
}
```

Generate speech using only the uploaded `voice` name:

```bash
curl http://127.0.0.1:8000/v1/audio/speech \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/home/admin/model",
    "input": "Hello, this is a Ming TTS uploaded embedding voice-only test after the fix.",
    "voice": "upload_emb_voice_verify",
    "response_format": "wav"
  }' \
  --output /home/ming_uploaded_emb_voice_only_fixed.wav
```

Result:

```text
HTTP 200
duration=6.72s, rms=0.0529, peak=0.392
```

- Negative test with invalid 191-dim direct speaker embedding:

```bash
python -c '
import json
json.dump([0.0] * 191, open("/tmp/ming_bad_191_embedding.json", "w"))
'

curl -s -o /tmp/ming_bad_191_upload_response.json -w "%{http_code}" \
  http://127.0.0.1:8000/v1/audio/voices \
  -F "name=ming_bad_191_verify" \
  -F "consent=local_smoke_test" \
  -F "speaker_embedding=</tmp/ming_bad_191_embedding.json"

cat /tmp/ming_bad_191_upload_response.json
```

Result:

```text
400
```

```json
{
  "error": {
    "message": "Ming speaker embedding must have 192 dims, got 191",
    "type": "BadRequestError",
    "param": null,
    "code": 400
  }
}
```

The invalid uploaded voice was not persisted.

- Regression check for existing uploaded audio/ref_text voice path:

Generate reference audio:

```bash
curl http://127.0.0.1:8000/v1/audio/speech \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/home/admin/model",
    "input": "Hello, this is a raw audio streaming test.",
    "response_format": "wav"
  }' \
  --output /home/ming_stream_main.wav
```

Upload voice with `audio_sample` and `ref_text`:

```bash
curl http://127.0.0.1:8000/v1/audio/voices \
  -F "name=upload_voice_main_ref_text" \
  -F "consent=local_smoke_test" \
  -F "audio_sample=@/home/ming_stream_main.wav;type=audio/wav" \
  -F "ref_text=Hello, this is a raw audio streaming test."
```

Generate speech using only the uploaded `voice` name:

```bash
curl http://127.0.0.1:8000/v1/audio/speech \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/home/admin/model",
    "input": "Hello, this is a regression test for uploaded audio voice.",
    "voice": "upload_voice_main_ref_text",
    "response_format": "wav"
  }' \
  --output /home/ming_uploaded_audio_voice_regression.wav
```

Result:

```text
HTTP 200
duration=6.08s, rms=0.118, peak=0.981
```

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py [hunk 1]
```diff
@@ -25,9 +25,18 @@ async def build(
         ref_audio_source = request.ref_audio
         voice_lower = request.voice.lower() if isinstance(request.voice, str) else None
         if ref_audio_source is None and voice_lower in server.uploaded_speakers:
-            ref_audio_source = server._get_uploaded_audio_data(request.voice)
-            if request.ref_text is None:
-                request.ref_text = server.uploaded_speakers[voice_lower].get("ref_text")
+            speaker_info = server.uploaded_speakers[voice_lower]
+            if speaker_info.get("embedding_source") == "direct":
+                if request.speaker_embedding is None:
+                    request.speaker_embedding = server._get_uploaded_speaker_embedding(request.voice)
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py:31]: Please reject non-192-dimensional Ming embeddings at upload time instead of persisting them. upload_voice_embedding() currently only logs a warning, returns success, and can replace an existing valid voice; this path then fails later during prompt assembly.

File: vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py [hunk 2]
```diff
@@ -25,9 +25,18 @@ async def build(
         ref_audio_source = request.ref_audio
         voice_lower = request.voice.lower() if isinstance(request.voice, str) else None
         if ref_audio_source is None and voice_lower in server.uploaded_speakers:
-            ref_audio_source = server._get_uploaded_audio_data(request.voice)
-            if request.ref_text is None:
-                request.ref_text = server.uploaded_speakers[voice_lower].get("ref_text")
+            speaker_info = server.uploaded_speakers[voice_lower]
+            if speaker_info.get("embedding_source") == "direct":
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py:29]: Please add a GPU-free regression test for this branch. A direct uploaded voice should call _get_uploaded_speaker_embedding(), avoid _get_uploaded_audio_data(), and propagate the 192-dimensional vector into prompt construction; the uploaded-audio path should remain covered too.

File: vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py (hunk 1 above)
Comment (LHXuuu) [reply vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py:31]: Done in 97f985d.

**Review by linyueqian — APPROVED** [review]
Approving — correct, well-scoped fix, verified end-to-end.

## Review

Traced the full data path: `_get_uploaded_speaker_embedding()` returns a `list[float]` (with path-traversal protection + a `direct`-only guard), which flows through `coerce_speaker_embeddings()` (`audio_prep.py`) where the 192-dim check is enforced and surfaced as HTTP 400. `use_zero_spk_emb` correctly resolves to `False` when the embedding is present, so it's actually consumed rather than zeroed. The branch is consistent with the existing `embedding_source == "direct"` handling in `voxcpm2.py` / `_apply_uploaded_speaker`, and it adds a defensive missing-audio guard to the `else` path.

## End-to-end verification

Applied the diff on a clean base and ran the full test plan against a live server (`inclusionAI/Ming-omni-tts-0.5B`, vLLM 0.24.0):

| Step | Result |
|------|--------|
| Extract 192-dim embedding | dims = 192 |
| Upload direct embedding | `embedding_source: direct, embedding_dim: 192` |
| **Generate by voice name only (the fix)** | **200**, dur 6.08s, rms 0.114 |
| **Negative: 191-dim embedding** | **400** `Ming speaker embedding must have 192 dims, got 191` |
| Regression: audio + ref_text path | 200, dur 6.08s |

6/6 passed. The 191-dim → 400 rejection is the key signal: it can only happen if the uploaded direct embedding is genuinely loaded and validated — before this fix that voice fell through to `use_zero_spk_emb` and the embedding was ignored.

## Optional (non-blocking)

A small unit test asserting the `embedding_source == "direct"` branch sets `request.speaker_embedding` would lock in the regression.

File: vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py (hunk 2 above)
Comment (LHXuuu) [reply vllm_omni/entrypoints/openai/tts_adapters/ming_tts.py:29]: Done in [9163d1d](https://github.com/vllm-project/vllm-omni/pull/5005/commits/9163d1d56e1822e26a403d82484c0ceaee941b54). Added three GPU-free regression tests:

- `test_upload_ming_voice_embedding_wrong_dims_rejected_without_replacing_existing`: verifies Ming-TTS rejects non-192-dim uploaded speaker embeddings before storing them, and does not replace an existing valid voice.
- `test_ming_adapter_uploaded_direct_embedding_uses_embedding_not_audio`: verifies direct uploaded voices call `_get_uploaded_speaker_embedding()`, skip `_get_uploaded_audio_data()`, and propagate the 192-dim vector into Ming prompt construction.
- `test_ming_adapter_uploaded_audio_path_preserves_ref_text`: verifies the existing uploaded audio/ref_text voice path still resolves audio, preserves `ref_text`, and builds the Ming prompt as before.

**Comment by LHXuuu** [comment]
> Approving — correct, well-scoped fix, verified end-to-end.
> 
> ## Review
> Traced the full data path: `_get_uploaded_speaker_embedding()` returns a `list[float]` (with path-traversal protection + a `direct`-only guard), which flows through `coerce_speaker_embeddings()` (`audio_prep.py`) where the 192-dim check is enforced and surfaced as HTTP 400. `use_zero_spk_emb` correctly resolves to `False` when the embedding is present, so it's actually consumed rather than zeroed. The branch is consistent with the existing `embedding_source == "direct"` handling in `voxcpm2.py` / `_apply_uploaded_speaker`, and it adds a defensive missing-audio guard to the `else` path.
> 
> ## End-to-end verification
> Applied the diff on a clean base and ran the full test plan against a live server (`inclusionAI/Ming-omni-tts-0.5B`, vLLM 0.24.0):
> 
> Step	Result
> Extract 192-dim embedding	dims = 192
> Upload direct embedding	`embedding_source: direct, embedding_dim: 192`
> **Generate by voice name only (the fix)**	**200**, dur 6.08s, rms 0.114
> **Negative: 191-dim embedding**	**400** `Ming speaker embedding must have 192 dims, got 191`
> Regression: audio + ref_text path	200, dur 6.08s
> 6/6 passed. The 191-dim → 400 rejection is the key signal: it can only happen if the uploaded direct embedding is genuinely loaded and validated — before this fix that voice fell through to `use_zero_spk_emb` and the embedding was ignored.
> 
> ## Optional (non-blocking)
> A small unit test asserting the `embedding_source == "direct"` branch sets `request.speaker_embedding` would lock in the regression.

Thanks for the review. I added GPU-free regression coverage for this: the direct uploaded voice path now asserts `_get_uploaded_speaker_embedding()` is used, `_get_uploaded_audio_data()` is skipped, and the 192-dim vector is propagated into `request.speaker_embedding`. I also kept coverage for the uploaded audio/ref_text path and the 191-dim upload rejection case.
